### PR TITLE
Fix prelu_backward TensorIterator split

### PR DESCRIPTION
**Pull request by zasdfgbnm**
We should have
```C++
    for (auto& sub_iter : iter.with_32bit_indexing()) {
      launch_prelu_cuda_backward_share_weights_kernel(sub_iter, weight_data);
    }
```

But I mistakenly wrote it as 

```C++
    for (auto& sub_iter : iter.with_32bit_indexing()) {
      launch_prelu_cuda_backward_share_weights_kernel(iter, weight_data);
    }
```

in my previous PR. Which leads to infinite recursion on it.

I found this bug when working on https://github.com/pytorch/pytorch/pull/34004

I also add a `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` to test for this.

Besides, the caller is already guaranteed contiguous, so we don't need to handle no-contiguous tensors.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Wait, please hold on for now, seems the testing for prelu backward is missing some cases, I will add them later in this PR.

**Comment by ailzhang** [comment]
@zasdfgbnm If you rebase on top of pytorch/master the XLA failure should be gone as well. Thanks!

File: aten/src/ATen/native/cuda/Activation.cu [hunk 1]
```diff
@@ -189,20 +191,11 @@ void launch_prelu_cuda_backward_share_weights_kernel(TensorIterator &iter, const
...
-  } else {
-    std::array<const int64_t*, 2> out_strides = {iter.strides(0).data(), iter.strides(1).data()};
-    std::array<const int64_t*, 2> inp_strides = {iter.strides(2).data(), iter.strides(3).data()};
-    prelu_cuda_backward_share_weights_kernel<scalar_t><<<grid, num_threads, 0, stream>>>(
-      numel, input_data, grad_out_data, input_grad_data, weight_grad_collector_data, weight_data,
-      OffsetCalculator<2>(iter.ndim(), iter.shape().data(), inp_strides.data()),
-      OffsetCalculator<2>(iter.ndim(), iter.shape().data(), out_strides.data())
-    );
-  }
+  TORCH_INTERNAL_ASSERT(iter.is_contiguous());
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Activation.cu:194]: Caller of `launch_prelu_cuda_backward_share_weights_kernel` already guarantees contiguous.

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin This PR is now ready. This is a bug fix for my previous prelu `CUDA_tensor_apply4` PR.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**